vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,521 @@
+            file_write_guid: LittleEndian::read_u128(&buffer[16..32]),
+            data_write_guid: LittleEndian::read_u128(&buffer[32..48]),
+            log_guid: LittleEndian::read_u128(&buffer[48..64]),
+            log_version: LittleEndian::read_u16(&buffer[64..66]),
+            version: LittleEndian::read_u16(&buffer[66..68]),
+            log_length: LittleEndian::read_u32(&buffer[68..72]),
+            log_offset: LittleEndian::read_u64(&buffer[72..80]),
+        })
+    }
+
+    /// Converts the header structure in to a buffer
+    fn get_header_as_buffer(&self, buffer: &mut [u8; HEADER_SIZE as usize]) {
+        LittleEndian::write_u32(&mut buffer[0..4], self.signature);
+        LittleEndian::write_u32(&mut buffer[4..8], self.checksum);
+        LittleEndian::write_u64(&mut buffer[8..16], self.sequence_number);
+        LittleEndian::write_u128(&mut buffer[16..32], self.file_write_guid);
+        LittleEndian::write_u128(&mut buffer[32..48], self.data_write_guid);
+        LittleEndian::write_u128(&mut buffer[48..64], self.log_guid);
+        LittleEndian::write_u16(&mut buffer[64..66], self.log_version);
+        LittleEndian::write_u16(&mut buffer[66..68], self.version);
+        LittleEndian::write_u32(&mut buffer[68..72], self.log_length);
+        LittleEndian::write_u64(&mut buffer[72..80], self.log_offset);
+    }
+
+    /// Creates and returns new updated header from the provided current header
+    pub fn update_header(
+        f: &mut File,
+        current_header: Header,
+        change_data_guid: bool,
+        mut file_write_guid: u128,
+        start: u64,
+    ) -> Result<Header> {
+        let mut buffer = [0u8; HEADER_SIZE as usize];
+        let mut data_write_guid = current_header.data_write_guid;
+
+        if change_data_guid {
+            data_write_guid = Uuid::new_v4().as_u128();
+        }
+
+        if file_write_guid == 0 {
+            file_write_guid = current_header.file_write_guid;
+        }
+
+        let mut new_header = Header {
+            signature: current_header.signature,
+            checksum: 0,
+            sequence_number: current_header.sequence_number + 1,
+            file_write_guid,
+            data_write_guid,
+            log_guid: current_header.log_guid,
+            log_version: current_header.log_version,
+            version: current_header.version,
+            log_length: current_header.log_length,
+            log_offset: current_header.log_offset,
+        };
+
+        new_header.get_header_as_buffer(&mut buffer);
+        new_header.checksum = crc32c::crc32c(&buffer);
+        new_header.get_header_as_buffer(&mut buffer);
+
+        f.seek(SeekFrom::Start(start)).map_err(Error::ReadHeader)?;
+        f.write(&buffer).map_err(Error::WriteHeader)?;
+
+        Ok(new_header)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct RegionTableHeader {
+    pub signature: u32,
+    pub checksum: u32,
+    pub entry_count: u32,
+    pub reserved: u32,
+}
```

Comment:
Ditto.

Reply:
According to the spec, reserved fields MUST be zero. So, we need to check them to see if they are consistent with the spec.